app/src/test/groovy/pl/futurecollars/invoicing/controller/InvoiceControllerTest.groovy

```diff
@@ -0,0 +1,146 @@
+                .contentAsString
+
+        and:
+        result == "[]"
+    }
+
+    def 'should not return  list when wrong data is sent'() {
+        given:
+        def url = "/invoic2es"
+        expect:
+        def result = mockMvc.perform(get(url))
+                .andExpect(status().isNotFound())
+    }
+
+    def "should return all invoices"() {
+        given:
+        def numberOfInvoices = 3
+        (1..numberOfInvoices).collect { id ->
+            def invoice = invoice(id)
+            invoice.id = inMemoryDataBase.save(invoice)
+        }
+        when:
+        def result = mockMvc.perform(get("/invoices"))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        then:
+        def resultList = jsonService.toObject(result, Invoice[])
+        resultList.size() == numberOfInvoices
+
+    }
+
+    def "should save invoice"() {
+        given:
+        Invoice invoice = invoice(1)
+        def url = "/invoice"
+
+        def invoiceJson = jsonService.toJson(invoice)
+        when:
+        def result = mockMvc.perform(post(url).content(invoiceJson)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        then:
+        result == "1"
+        def invoiceOptional = inMemoryDataBase.getById(1)
+        invoiceOptional.isPresent()
+    }
+
+    def "should not save invoice when wrong data is sent"() {
+        given:
+        def url = "/invoice"
+
+        expect:
+        mockMvc.perform(post(url).content("siema")
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+    }
+
+    def "should update invoice"() {
+        given:
+        def updateInvoice = originalInvoice
+        def url = "/invoice/{}"
+
+        def invoiceAsJson = jsonService.toJson(updateInvoice)
+        when:
+        def result = mockMvc.perform(put(url ).content(invoiceAsJson)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+        then:
+        updateInvoice.date == originalInvoice.date
+    }
+
+    def "should not update invoice when wrong data is sent"() {
+        given:
+        Invoice updateInvoice = invoice(1)
+        def url = "/inv2ice/{}"
+        updateInvoice.id = inMemoryDataBase.save(updateInvoice)
+        def invoiceAsJson = jsonService.toJson(updateInvoice)
+        expect:
+        def result = mockMvc.perform(put(url).content(invoiceAsJson)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNotFound())
+    }
+
+    def "invoice can be deleted"() {
+        given:
+        def url = "/invoice/{}"
+        expect:
+        mockMvc.perform(MockMvcRequestBuilders.delete(url))
+                .andExpect(status().isBadRequest())
+    }
```

Comment:
tutaj to samo, test ma być pozytywną ścieżką usuwania invoice, a sprawdzasz bad request. Powinno to wyglądać tak, że dodajesz coś do bazy, potem to usuwasz endpointem i w then sprawdzasz czy baza jest pusta. a status to powinno być ok